runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+		return err
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	info, err := os.Stat(dst)
+	switch {
+	case err == nil:
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	case os.IsNotExist(err):
+		// this occurs when a full path is provided including the file name
+	default:
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+var (
+	stdout *os.File
+	stderr *os.File
+)
+
+func init() {
+	var err error
+	stdout, err = os.OpenFile("/var/log/firecracker-containerd-test/stdout", os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		panic(err)
+	}
+	stderr, err = os.OpenFile("/var/log/firecracker-containerd-test/stderr", os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (j runcJailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.Command(j.runcBinaryPath, "run", containerName)
+	cmd.Dir = j.OCIBundlePath()
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	return cmd
+}
+
+// overwriteConfig will set the proper default values if a field had not been set.
+func (j runcJailer) overwriteConfig(cfg *Config, socketPath, configPath string) error {
+	spec := specs.Spec{}
+	configBytes, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(configBytes, &spec); err != nil {
+		return err
+	}
+
+	if spec.Process.User.UID != 0 ||
+		spec.Process.User.GID != 0 {
+		return fmt.Errorf(
+			"using UID %d and GID %d, these values must not be set",
+			spec.Process.User.UID,
+			spec.Process.User.GID,
+		)
+	}
+
+	spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+
+	spec.Root.Path = rootfsFolder
+	spec.Root.Readonly = false
+	spec.Process.User.UID = j.uid
+	spec.Process.User.GID = j.gid
+
+	configBytes, err = json.Marshal(&spec)
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(configPath, configBytes, 0444); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setDefaultConfigValues will process the spec file provided and allow any
+// empty/zero values to be replaced with default values.
+func (j runcJailer) setDefaultConfigValues(cfg *Config, socketPath string, spec specs.Spec) specs.Spec {
+	if spec.Process != nil {
+		if spec.Process.Args == nil {
+			cmd := firecracker.VMCommandBuilder{}.
```

Comment:
I'm confused: why does this only happen if `spec.Process != nil && spec.ProcessArgs == nil`?

Reply:
Ah, we also need to check if process is `nil` all together. So basically if the user provides no process section in the runc config, we will populate a simple one with only command arguments. If a process section was specified but no process args, then we will also populate the command to be ran.